bve-rex/src/game/mod.rs

```diff
@@ -0,0 +1,113 @@
+use bve_rex_sys::*;
+use std::ffi::c_void;
+use std::ffi::CString;
+use std::mem::transmute;
+use std::os::raw::c_char;
+
+pub trait GameBuilder: Sized {
+    type Built: Game;
+
+    fn build(&mut self, interface: &mut rx::render::frontend::interface) -> Box<Self::Built>;
+}
+
+pub trait Game: Sized {
+    fn on_init(&mut self) -> bool;
+    fn on_slice(&mut self, input: &mut rx::input::input) -> rx::game_status;
+    fn on_resize(&mut self, dimensions: &rx::math::vec2z);
+}
+
+// This doesn't need to be extern C nor c_interface(nomangle) as I am calling it from create, which is c_interface
+// itself
+unsafe fn build_wrapper<B>(data: *mut c_void, interface: &mut rx::render::frontend::interface) -> *mut bve::game
+where
+    B: GameBuilder,
+{
+    // This data pointer comes from the start function.
+    // This pointer isn't used again, so we are responsible for destroying this.
+    let mut data = Box::from_raw(data as *mut B);
+
+    // Once we have the pointer to the builder, call the builder function
+    let built: Box<B::Built> = data.build(interface);
+    let game = bve::game::new(
+        // It will be dropped by the `drop_game` function.
+        Box::into_raw(built) as *mut c_void,
+        Some(on_init_wrapper::<B::Built>),
+        Some(on_slice_wrapper::<B::Built>),
+        Some(on_resize_wrapper::<B::Built>),
+        Some(drop_game::<B::Built>),
+    );
+    // Box the game wrapper up and turn it into a pointer.
+    // This is dropped in the `destroy` hook provided to rex.
+    Box::into_raw(Box::new(game))
```

Comment:
Why do you make a Box from the game pointer, only to immediately un-box it?

Reply:
that's the translation from `bve::Game` -> `Box<bve::game>` to `*mut bve::game`

Reply:
Update: I get it